### PR TITLE
Add CopyToClipboardButton component

### PR DESCRIPTION
**Pull request by bernardodiasc**
**Release Type:** *Non-breaking feature*

Fixes https://x-team-internal.atlassian.net/browse/XP-3335

## Description

Add CopyToClipboardButton component.

**Note:** Please merge https://github.com/x-team/xp-ui/pull/443 before this one.

## Checklist

- [x] set yourself as an assignee
- [x] set appropriate labels for a PR (`In Review` or `In Progress` depending on its status)
- [x] move respective JIRA issue to the `IN REVIEW` column
- [x] make sure your code lints (`npm run lint`)
- [x] Flow typechecks passed (`npm run flow`)
- [x] Snapshots tests passed (`npm run jest`)
- [x] component's documentation (`.stories.js` file) is changed or added accordingly to reflect any new or updated use cases or variants usage
- [x] if any snapshots have been changed, verify that component still works and looks as expected and update the changed snapshot
- [x] **manually tested the app** by running it in several different browsers (Firefox, Chrome, Opera, Safari, MS IE/Edge, etc.) and checked nothing is broken and operates as expected!

## Related PRs

branch | PR
------ | ------
`XP-2894-upgrade-shortcuts-with-updated-genericcopytoclipboard` | [XP UI](https://github.com/x-team/xp-ui/pull/443)
`XP-3335-upgrade-fullname-with-updated-genericcopytoclipboard` | [XP](https://github.com/x-team/xp/pull/2000)

## Steps to Test or Reproduce

1. Check the component at http://localhost:9001/?path=/story/ui-components-copytoclipboardbutton--basic-usage

## Impacted Areas in Application

New `CopyToClipboardButton` included.

### CODE REVIEW

File: src/components/ui/CopyToClipboardButton.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    hover: 'default'
+  }
+
+  render () {
+    const { text, visible, color, hover } = this.props
+    const visibleClassName = visible ? cx.copyButtonVisible : ''
+
+    return text ? (
+      <span role='button' className={`${cx.copyButton} ${visibleClassName}`}>
+        <GenericCopyToClipboard text={text} tooltipXOffset={-43}>
```
Comment (nicksp) [inline src/components/ui/CopyToClipboardButton.js]: Let's make it component reusable so we can integrate it in other places as well. This will mean dynamically specifying `tooltipXOffset` per use case.

File: src/components/ui/GenericCopyToClipboard/Tooltip.js [hunk 2]
```diff
@@ -60,12 +68,13 @@ const cx = {
...
-  copied: boolean
+  copied: boolean,
+  tooltipXOffset: number
 }
 
-const Tooltip = ({ copied }: Props) => {
+const Tooltip = ({ copied, tooltipXOffset }: Props) => {
   return (
-    <span className={cx.container}>
+    <span className={cx.container} style={{ left: `${tooltipXOffset}px` }}>
```
Comment (nicksp) [inline src/components/ui/GenericCopyToClipboard/Tooltip.js]: Let's extract `style` prop into a variable due to the same reason I explained in another PR.

File: src/components/ui/CopyToClipboardButton.js (hunk 1 above)
Comment (bernardodiasc) [reply src/components/ui/CopyToClipboardButton.js]: Sorry, I didn't get what you meant. Can you be more specific about what changes you expect?

Comment (nicksp) [reply src/components/ui/CopyToClipboardButton.js]: Let's use this new component in other places as well: shortcuts, email.

File: src/components/ui/GenericCopyToClipboard/Tooltip.js (hunk 2 above)
Comment (bernardodiasc) [reply src/components/ui/GenericCopyToClipboard/Tooltip.js]: I'll change this in the scope of https://github.com/x-team/xp-ui/pull/443, notice that only this commit https://github.com/x-team/xp-ui/pull/444/commits/9b90e7ae296411e235d0c9640d541b40537f9b23 is for this PR, the others is from the https://github.com/x-team/xp-ui/pull/443 :)

Comment (nicksp) [reply src/components/ui/GenericCopyToClipboard/Tooltip.js]: whatever you prefer, just mentioned for the sake of review ;-)

File: src/components/ui/CopyToClipboardButton.js (hunk 1 above)
Comment (bernardodiasc) [reply src/components/ui/CopyToClipboardButton.js]: I commented about that here https://github.com/x-team/xp-ui/pull/443#issuecomment-533485390, Shortcuts needs a unique implementation because of styles. For email, that's already being used, you chan see here https://github.com/x-team/xp/pull/2001

File: src/components/ui/GenericCopyToClipboard/Tooltip.js (hunk 2 above)
Comment (bernardodiasc) [reply src/components/ui/GenericCopyToClipboard/Tooltip.js]: No problem, just wanted to make sure that you are aware of the branches overlapping :)

File: src/components/ui/CopyToClipboardButton.js (hunk 1 above)
Comment (bernardodiasc) [reply src/components/ui/CopyToClipboardButton.js]: That's the problem when design thinking doesn't work together with development team. This extra abstraction would only work if designers guarantee both places are in fact the same. And we can tell they are not just looking that how they are implemented on each places.

We already have `GenericCopyToClipboard` and `SvgIcon` abstractions in place, creating an abstraction on top of that which is flexible enough to handle different implementations need to be made carefully.

> Will providing styles when calling this component solve the problem you mentioned?

It's possible to do, of course, but this doesn't seem beneficial approach. The only thing they have in common is `GenericCopyToClipboard` and `SvgIcon`, which are already abstractions, but not even the values used on these components are the same in `Shortcuts` and `CopyToClipboardButton`. As far as I read in many different sources, scalable development doesn't necessarily means abstract all the things, because sometimes this approach leads to entangled codebase.

I'll make another try, but as far as I can tell it will make `CopyToClipboardButton` unnecessarily complex. Or maybe this component should be renamed to `ProfileFormCopyToClipboardButton` to make it explicit that is not a generic thing. `Shortcuts` is not a `ProfileForm` element.

Well, this idea doesn't sound good to me for all the reasons above but let me try again, will get back with something.

Comment (nicksp) [reply src/components/ui/CopyToClipboardButton.js]: I understand. Why did we introduce this new abstraction then if we can't reuse it everywhere? Consider removing if you wish, so we use copy to clipboard consistently.

Comment (bernardodiasc) [reply src/components/ui/CopyToClipboardButton.js]: > Why did we introduce this new abstraction then if we can't reuse it everywhere?

It supposed to be reused on all ProfileForm elements, along with PencilButton. It is in fact as you can see in this PR https://github.com/x-team/xp/pull/2001. The Shortcuts is the only exception that can't reuse that.

**Comment by bernardodiasc** [comment]
Hey @nicksp, if you agree with this comment https://github.com/x-team/xp-ui/pull/444#discussion_r326574598, then this PR should be ready. I gave a second try on how Shortcuts could use CopyToClipboardButton. Unfortunately the implementation on these places are different. Making CopyToClipboardButton flexible enough for Shortcuts will not help much and it will raise unnecessary complexity. I know the name "CopyToClipboardButton" make it look generic, but as the "PencilButton", they are intended for ProfileForm elements and doesn't seem appropriate to force it to be generic for every CopyToClipboard usages. Please let me know what you think about.

**Comment by bernardodiasc** [comment]
I resolved conflicts on this PR, should be ready for review.

File: src/components/ui/CopyToClipboardButton.stories.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import React from 'react'
```
Comment (nicksp) [inline src/components/ui/CopyToClipboardButton.stories.js]: As agreed before, let's remember to add Flow typings check to story files as well.

File: src/components/ui/PencilButton.stories.js [hunk 4]
```diff
@@ -3,27 +3,32 @@ import { storiesOf } from '@storybook/react'
 
 import PencilButton from './PencilButton'
 
+const Body = ({ children, background }) => (
```
Comment (nicksp) [inline src/components/ui/PencilButton.stories.js]: Let's add `// @flow` here as well.

File: src/components/ui/Shortcuts.stories.js [hunk 5]
```diff
@@ -18,7 +18,7 @@ const Body = ({ children }) => (
 
 const seedCount = 7
 const links = Array(seedCount).fill({}).reduce((acc, cur) => {
```
Comment (nicksp) [inline src/components/ui/Shortcuts.stories.js]: `// @flow`

Comment (bernardodiasc) [reply src/components/ui/Shortcuts.stories.js]: this one already have flow 👍
